### PR TITLE
support keras 2.3 and tensorflow 2.0 

### LINKED ISSUE #789: Tensorflow2.0目前是否全面支持？
如题，我目前的运行环境使用是tf2.0版本，keras是为2.3.0。
但无法执行
报错信息如下：
```
~/anaconda3/lib/python3.7/site-packages/keras/engine/training.py in _prepare_total_loss(self, masks)
    690 
    691                     output_loss = loss_fn(
--> 692                         y_true, y_pred, sample_weight=sample_weight)
    693 
    694                 if len(self.outputs) > 1:

TypeError: __call__() got an unexpected keyword argument 'sample_weight'
```

### PR DESCRIPTION
**Pull request by matthew-z**
- update requirements.txt: keras=2.3.0 and  tensorflow >= 2.0.0
- upgrade pip in .travis.yml (tf 2.0 requires pip >= 19)
- make raking losses inherit keras.losses.Loss to support `sample_weight` keyword param
- replace some `keras.backend.tf` with `tf` (`K.tf` does not exist anymore in 2.3.0 as keras is going to be synced with tf.keras and drop multi-backend)
- add `clear_session` before `prepare` in model tests to prevent OOM during CI test

fix #789

### CODE REVIEW

**Comment by matthew-z** [comment]
It seems that the `DecayingDropoutLayer` is broken in `tf 2.0`: replacing it with normal dropout could fix the error

**Comment by uduse** [comment]
Nice work! I just added you as one of the maintainers, so once this PR is finished and approved by at least 2 reviewers, you may go ahead and merge this.

**Comment by bwanglzu** [comment]
I was wondering does it still work for the ppl using tensorflow 1.x.x?

**Comment by bwanglzu** [comment]
@matthew-z also do not forget to add dependencies in `setup.py`

**Comment by matthew-z** [comment]
OK I have changed dependencies in setup.py as well.

It works with Keras 2.3 + tf 1.14, but not with Keras 2.2 + tf 1.14. I have not tested older tf versions.

Also, I think we should consider fully using `tf.keras` in tf 2.0. 

> Going forward, we recommend that users consider switching their Keras code to tf.keras in TensorFlow 2.0. It implements the same Keras 2.3.0 API (so switching should be as easy as changing the Keras import statements), but it has many advantages for TensorFlow users, such as support for eager execution, distribution, TPU training, and generally far better integration between low-level TensorFlow and high-level concepts like Layer and Model. It is also better maintained.
https://github.com/keras-team/keras/releases

**Comment by bwanglzu** [comment]
@ZizhenWang  did some job in #643 , can you review @ZizhenWang ?

File: matchzoo/losses/rank_cross_entropy_loss.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 """The rank cross entropy loss."""
-import numpy as np
 
 from keras import layers, backend as K
+import numpy as np
```
Comment (bwanglzu) [inline matchzoo/losses/rank_cross_entropy_loss.py]: can put numpy above `keras` import?

File: matchzoo/layers/matching_layer.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 """An implementation of Matching Layer."""
 import typing
 
+import tensorflow as tf
 from keras import backend as K
```
Comment (bwanglzu) [inline matchzoo/layers/matching_layer.py]: take out K compeletely?

such as 
```python
# line78
tf.match.l2_normalize
```

File: matchzoo/losses/rank_cross_entropy_loss.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 """The rank cross entropy loss."""
-import numpy as np
 
 from keras import layers, backend as K
```
Comment (bwanglzu) [inline matchzoo/losses/rank_cross_entropy_loss.py]: same

File: matchzoo/losses/rank_hinge_loss.py [hunk 4]
```diff
@@ -28,13 +30,18 @@ def __init__(self, num_neg: int = 1, margin: float = 1.0):
...
         self._num_neg = num_neg
         self._margin = margin
 
-    def __call__(self, y_true: np.array, y_pred: np.array) -> np.array:
+    def call(self, y_true: np.array, y_pred: np.array,
+             sample_weight=None) -> np.array:
         """
         Calculate rank hinge loss.
 
+        :param **kwargs:
```
Comment (bwanglzu) [inline matchzoo/losses/rank_hinge_loss.py]: where is `**kwargs`?

File: matchzoo/losses/rank_cross_entropy_loss.py [hunk 5]
```diff
@@ -48,7 +53,10 @@ def __call__(self, y_true: np.array, y_pred: np.array) -> np.array:
             labels.append(neg_labels)
         logits = K.concatenate(logits, axis=-1)
         labels = K.concatenate(labels, axis=-1)
-        return -K.mean(K.sum(labels * K.log(K.softmax(logits) + np.finfo(float).eps), axis=-1))
+        smoothed_prob = K.softmax(logits) + np.finfo(float).eps
```
Comment (bwanglzu) [inline matchzoo/losses/rank_cross_entropy_loss.py]: differrence between `tf.nn.softmax` and `K.softmax`?

File: matchzoo/models/drmm.py [hunk 6]
```diff
@@ -114,7 +114,7 @@ def attention_layer(cls, attention_input: typing.Any,
             )(dense_input)
         # shape = [B, L, 1]
         attention_probs = keras.layers.Lambda(
-            lambda x: keras.layers.activations.softmax(x, axis=1),
+            lambda x: keras.activations.softmax(x, axis=1),
```
Comment (bwanglzu) [inline matchzoo/models/drmm.py]: same

File: matchzoo/models/drmmtks.py [hunk 7]
```diff
@@ -127,7 +127,7 @@ def attention_layer(cls, attention_input: typing.Any,
             )(dense_input)
         # shape = [B, L, 1]
         attention_probs = keras.layers.Lambda(
-            lambda x: keras.layers.activations.softmax(x, axis=1),
+            lambda x: keras.activations.softmax(x, axis=1),
```
Comment (bwanglzu) [inline matchzoo/models/drmmtks.py]: same

File: matchzoo/models/knrm.py [hunk 8]
```diff
@@ -69,13 +70,13 @@ def build(self):
                 mu = 1.0
             mm_exp = self._kernel_layer(mu, sigma)(mm)
             mm_doc_sum = keras.layers.Lambda(
-                lambda x: K.tf.reduce_sum(x, 2))(mm_exp)
-            mm_log = keras.layers.Activation(K.tf.log1p)(mm_doc_sum)
+                lambda x: K.sum(x, 2))(mm_exp)
+            mm_log = keras.layers.Activation(tf.math.log1p)(mm_doc_sum)
             mm_sum = keras.layers.Lambda(
-                lambda x: K.tf.reduce_sum(x, 1))(mm_log)
+                lambda x: K.sum(x, 1))(mm_log)
```
Comment (bwanglzu) [inline matchzoo/models/knrm.py]: `tf.math.reduce_sum`?

**Comment by matthew-z** [comment]
As keras 2.3 removed `K.tf.*`, so my initial thought is simply to fix `K.tf` in the affected files. Some ops have alias in `K.*`, but some of them don't, So you think we should replace all `K.*` with `tf.*`?

**Comment by matthew-z** [comment]
> @ZizhenWang did some job in #643 , can you review @ZizhenWang ?
> 
> @matthew-z checkout description in #642

I was wondering why the PR was closed without merging?

File: README.md [hunk 9]
```diff
@@ -117,7 +117,7 @@ If you're interested in the cutting-edge research progress, please take a look a
 
 ## Install
 
-MatchZoo is dependent on [Keras](https://github.com/keras-team/keras), please install one of its backend engines: TensorFlow, Theano, or CNTK. We recommend the TensorFlow backend. Two ways to install MatchZoo:
+MatchZoo is dependent on [Keras](https://github.com/keras-team/keras) and [Tensorflow](https://github.com/tensorflow/tensorflow). Two ways to install MatchZoo:
```
Comment (bwanglzu) [inline README.md]: nice :thumbsup:

**Comment by bwanglzu** [comment]
@matthew-z I have no clue.. why they closed the previous pr.

**Comment by matthew-z** [comment]
Thank you for the review. I guess I need one more reviewer to approve before merging it?

**Comment by bwanglzu** [comment]
@matthew-z Yes! I'll send them a message.

**Comment by ZizhenWang** [comment]
Sorry for the late response, I will review the code ASAP.

File: matchzoo/contrib/layers/decaying_dropout_layer.py [hunk 10]
```diff
@@ -1,10 +1,9 @@
 """An implementation of Decaying Dropout Layer."""
-import typing
 
 import keras.backend as K
+import tensorflow as tf
```
Comment (ZizhenWang) [inline matchzoo/contrib/layers/decaying_dropout_layer.py]: a little advice, import same package in the same block, like
```
import tensorflow as tf
import keras.backend as K
from keras.engine import Layer
```

File: matchzoo/contrib/layers/semantic_composite_layer.py [hunk 11]
```diff
@@ -1,8 +1,7 @@
 """An implementation of EncodingModule for DIIN model."""
-import typing
 
-import keras
 import keras.backend as K
+import tensorflow as tf
```
Comment (ZizhenWang) [inline matchzoo/contrib/layers/semantic_composite_layer.py]: the import order

**Comment by ZizhenWang** [comment]
@matthew-z Nice work! You can merge PR after doing the little changes as I comment.

I notice that you use `tensorflow` explicitly rather than as one of the `keras` backends. In that way, `tf` seems parallel to `keras` in this repo, just as @bwanglzu mentioned.

At #643 I use `tf.keras` instead of `keras`, it makes `matchzoo` dependent entirely on `tensorflow`, but that PR is delayed by other things. If you have interest on `tf.keras`, we can open another PR and remove `keras`.

Thanks for your contribution again!

**Comment by uduse** [comment]
Good job! Time for version 2.3!

**Comment by matthew-z** [comment]
BTW, CI failed because the certificate of nlp.stanford.edu expired several hours ago....

**Pull request merged**